File: algo_8_math/L179-m.cpp
//
// Created by LiHeng on 2020/6/30.
//
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;
class Solution {
public:
    string largestNumber(vector<int>& nums) {
        string res;
        sort(nums.begin(), nums.end(), cmp);
        for (int i = 0; i < nums.size(); ++i) {
            res += to_string(nums[i]);
        }
        if (res[0] == '0') {
            res = '0';
        }
        return res;
    }

    static bool cmp(const int &a, const int &b) {
        string s1 = to_string(a) + to_string(b);
        string s2 = to_string(b) + to_string(a);
        return s1 > s2;
    }
};

// 自定义了一个比较函数，c++ stl sort 自定义比较函数必须是静态
// 对于2个数字a,b合在一起谁大 就看ab和ba谁大 如果ab>ba 那就认为排序时候a>b a排在前面

File: python-algorithm/algo_01_sort/L179-m.py
#--coding:utf-8--
# 179. 最大数
from functools import cmp_to_key
class Solution:
    def largestNumber(self, nums: List[int]) -> str:
        def cmp(m, n):
            s1 = str(m)+str(n)
            s2 = str(n)+str(m)
            if int(s1)>int(s2):
                return -1
            elif int(s1)<int(s2):
                return 1
            else:
                return 0
        # 从大到小排序
        nums.sort(key = cmp_to_key(cmp))
        res = "".join([str(it) for it in nums])
        # 处理[0, 0]特殊情况
        return  res if int(res) else "0"